### PR TITLE
Add test for parsing maps with integer keys

### LINKED ISSUE #25026: Map does not support Integer key very well since 21.4
Below test case worked before 21.4. It seems there's parsing issue with all Map columns using Integer as key.
```sql
set allow_experimental_map_type = 1;
CREATE TABLE IF NOT EXISTS test_maps(ma Map(Integer, Array(String)), mi Map(Integer, Integer), ms Map(String, String)) ENGINE = Memory;
-- Exception on client:
-- Code: 62. DB::Exception: Cannot parse expression of type Map(Int32,Array(String)) here: {1:['11','12'],2:['22','23']},{1:11,2:22},{'k1':'v1','k2':'v2'}): data for INSERT was parsed from query
insert into test_maps values ({1:['11','12'],2:['22','23']},{1:11,2:22},{'k1':'v1','k2':'v2'});
```

Just tried 21.6 and it's still the same:
```sql
:) create table if not exists test_map_with_int_key(m Map(UInt32, UInt32)) engine = Memory;
...
:) SELECT CAST(([1, 2], [11, 22]), 'Map(UInt32, UInt32)') AS m;

SELECT CAST(([1, 2], [11, 22]), 'Map(UInt32, UInt32)') AS m

Query id: dbe47185-bc24-488b-a85d-761616e3199d

┌─m───────────┐
│ {1:11,2:22} │
└─────────────┘
:) insert into test_map_with_int_key SELECT CAST(([1, 2], [11, 22]), 'Map(UInt32, UInt32)') as m

INSERT INTO test_map_with_int_key SELECT CAST(([1, 2], [11, 22]), 'Map(UInt32, UInt32)') AS m

Query id: 6877901f-1518-48f9-ba65-d54ed7326388

Ok.
:) insert into test_map_with_int_key values({1:11,2:22})

INSERT INTO test_map_with_in_key VALUES

Query id: 3f6cae42-064a-467f-baae-b7f323e910c0


0 rows in set. Elapsed: 0.004 sec.

Received exception from server (version 21.6.3):
Code: 60. DB::Exception: Received from localhost:9000. DB::Exception: Table default.test_map_with_in_key doesn't exist.
```

P.S. I'll submit PRs for integration test this week so that we can spot this kind of issue early.

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fixes #25026.

### CODE REVIEW

File: src/DataTypes/Serializations/SerializationMap.cpp [hunk 1]
```diff
@@ -177,11 +177,20 @@ void SerializationMap::serializeText(const IColumn & column, size_t row_num, Wri
 
 void SerializationMap::deserializeText(IColumn & column, ReadBuffer & istr, const FormatSettings & settings) const
 {
+    /// We need read int safe, because in unsafe version
```
Comment (kitaisreal) [inline src/DataTypes/Serializations/SerializationMap.cpp:180]: @alexey-milovidov maybe better to remove unsafe parsing and check performance. Does it really 1.5x faster ?

Comment (kitaisreal) [reply src/DataTypes/Serializations/SerializationMap.cpp:180]: @CurtizJ current solution seems to hacky need to discuss with @alexey-milovidov .

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/DataTypes/Serializations/SerializationNumber.cpp [hunk 2]
```diff
@@ -21,12 +21,17 @@ void SerializationNumber<T>::serializeText(const IColumn & column, size_t row_nu
...
 template <typename T>
-void SerializationNumber<T>::deserializeText(IColumn & column, ReadBuffer & istr, const FormatSettings &) const
+void SerializationNumber<T>::deserializeText(IColumn & column, ReadBuffer & istr, const FormatSettings & settings) const
 {
     T x;
 
     if constexpr (is_integer_v<T> && is_arithmetic_v<T>)
-        readIntTextUnsafe(x, istr);
+    {
+        if (settings.read_int_safe)
```
Comment (alexey-milovidov) [inline src/DataTypes/Serializations/SerializationNumber.cpp:30]: This branch alone can introduce some cost.

Comment (CurtizJ) [reply src/DataTypes/Serializations/SerializationNumber.cpp:30]: But it's already in virtual method. I think it's not such significant.

File: src/DataTypes/Serializations/SerializationMap.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/DataTypes/Serializations/SerializationMap.cpp:180]: @CurtizJ after https://github.com/ClickHouse/ClickHouse/pull/27155 merged, we can keep only tests.

**Comment by kitaisreal** [comment]
@CurtizJ please generate references for tests 01550_type_map_formats, 01318_map_add_map_subtract_on_map_type to fix formatting issues.

**Review by kitaisreal — APPROVED** [review]

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**